coderd/userauth.go

```diff
@@ -201,6 +203,231 @@ func (api *API) postConvertLoginType(rw http.ResponseWriter, r *http.Request) {
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	err = api.Database.InTx(func(tx database.Store) error {
+		//nolint:gocritic // In order to change a user's password, we need to get the user first - and can only do that in the system auth context.
+		user, err := tx.GetUserByEmailOrUsername(dbauthz.AsSystemRestricted(ctx), database.GetUserByEmailOrUsernameParams{
+			Email: req.Email,
+		})
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			logger.Error(ctx, "unable to fetch user by email", slog.F("email", req.Email), slog.Error(err))
+			return xerrors.Errorf("get user by email: %w", err)
+		}
+		aReq.Old = user
+
+		equal, err := userpassword.Compare(string(user.HashedOneTimePasscode), req.OneTimePasscode)
+		if err != nil {
+			logger.Error(ctx, "unable to compare one time passcode", slog.Error(err))
+			return xerrors.Errorf("compare one time passcode: %w", err)
+		}
+
+		now := dbtime.Now()
+		if !equal || now.After(user.OneTimePasscodeExpiresAt.Time) {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Incorrect email or one-time-passcode.",
+			})
+			return nil
+		}
+
+		if err := userpassword.Validate(req.Password); err != nil {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Invalid password.",
+				Validations: []codersdk.ValidationError{
+					{
+						Field:  "password",
+						Detail: err.Error(),
+					},
+				},
+			})
+			return nil
+		}
+
+		if equal, _ = userpassword.Compare(string(user.HashedPassword), req.Password); equal {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "New password cannot match old password.",
+			})
+			return nil
+		}
+
+		newHashedPassword, err := userpassword.Hash(req.Password)
+		if err != nil {
+			logger.Error(ctx, "unable to hash user's password", slog.Error(err))
+			return xerrors.Errorf("hash user password: %w", err)
+		}
+
+		//nolint:gocritic // We need the system auth context to be able to update the user's password.
+		err = tx.UpdateUserHashedPassword(dbauthz.AsSystemRestricted(ctx), database.UpdateUserHashedPasswordParams{
+			ID:             user.ID,
+			HashedPassword: []byte(newHashedPassword),
+		})
+		if err != nil {
+			logger.Error(ctx, "unable to delete user's hashed password", slog.Error(err))
+			return xerrors.Errorf("update user hashed password: %w", err)
+		}
+
+		//nolint:gocritic // We need the system auth context to be able to delete all API keys for the user.
+		err = tx.DeleteAPIKeysByUserID(dbauthz.AsSystemRestricted(ctx), user.ID)
+		if err != nil {
+			logger.Error(ctx, "unable to delete user's api keys", slog.Error(err))
+			return xerrors.Errorf("delete api keys for user: %w", err)
+		}
+
+		auditUser := user
+		auditUser.HashedPassword = []byte(newHashedPassword)
+		auditUser.OneTimePasscodeExpiresAt = sql.NullTime{}
+		auditUser.HashedOneTimePasscode = nil
+		aReq.New = auditUser
+
+		rw.WriteHeader(http.StatusOK)
```

Comment:
I will challenge @dannykopping here. This is a pure HTTP 204 case to me, see [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204).

Reply:
I disagree, since this is a `POST` endpoint.

> A 204 No Content in response to these request methods has the following meaning and results:
> - [DELETE](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE): The action was successful, and no further information needs to be supplied.
> - [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT): The action was successful, and the [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) value contains the entity tag for the new representation of that target resource.

Whereas for [`200 OK`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200):

> A 200 OK response has a different meaning and format depending on the HTTP request method. Here's how they vary for different methods:
> ...
> [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST): An action succeeded; the response has a message body describing the result.
> ...

Reply:
> [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST): An action succeeded; the response has a message body describing the result.

There is no response body, which implies HTTP 204, right?

Reply:
Also, following: [RFC 7231](https://www.rfc-editor.org/rfc/rfc7231#section-6.3.1)

>    Aside from responses to CONNECT, a 200 response always has a payload,
   though an origin server MAY generate a payload body of zero length.
   If no payload is desired, an origin server ought to send 204 (No
   Content) instead.  For CONNECT, no payload is allowed because the
   successful result is a tunnel, which begins immediately after the 200
   response header section.

Reply:
Ah, I see what you mean.

This could go either way: we could either add a response body like `password updated` with a `200 OK` or revert to a `204 No Content`. I don't feel strongly about either.

Whatever feels congruent with other similar endpoints will be the best IMHO; consistency is the thing we should be aiming for here I think, on reflection.

Reply:
Agree, up to @DanielleMaywood to pick the best way 👍

Reply:
I don't believe we have strict guidelines on this but since there's only one type of status that can be returned for success here, I think status no content is the most appropriate. Same as we return when we put a new password for a user.

Reply:
I've gone with 204 No Content 👍